File: app/static/js/fuel_banner.js
/* Banner grado combustible — logos oficiales (AVGAS rojo / JET negro). */
'use strict';

(function (global) {
  const AVGAS_BG = '#C0392B';
  const JET_BG = '#1C1C1C';

  const LOGO = {
    jet: {
      svg: '/static/img/fuel/jet-a1.svg',
      png: '/static/img/fuel/jet-a1-oficial.png',
      alt: 'JET A-1',
    },
    avgas: {
      svg: '/static/img/fuel/avgas-100ll.svg',
      png: '/static/img/fuel/avgas-100ll-oficial.png',
      alt: 'AVGAS 100 LL',
    },
  };

  function escapeHtml(s) {
    return String(s ?? '').replace(/[&<>"']/g, (c) => (
      { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]
    ));
  }

  function cleanTipo(tipo, forCliente) {
    const t = String(tipo || '').trim();
    if (!t) return '';
    if (forCliente && /^(S\/D|SD|N\/D|ND|-|—)$/i.test(t)) return '';
    return t;
  }

  function resolveKind(combustible) {
    const raw = String(combustible || '').trim();
    if (!raw) return { kind: '', label: '', colorCode: '' };
    const u = raw.toUpperCase().replace(/[_-]+/g, ' ').replace(/\s+/g, ' ').trim();
    const compact = u.replace(/\s+/g, '');
    if (compact.includes('JET') || u.includes('AEROKEROSENE')) {
      return { kind: 'jet', label: 'JET A-1', colorCode: 'NEGRO' };
    }
    if (u.includes('AVGAS') || compact.includes('100LL')) {
      return { kind: 'avgas', label: 'AVGAS 100LL', colorCode: 'ROJO' };
    }
    return { kind: 'unknown', label: raw, colorCode: '' };
  }

  function logoMarkup(kind, { chip } = {}) {
    const L = LOGO[kind];
    if (!L) return '';
    const cls = chip ? 'chip__logo' : 'fuel-banner__logo';
    const wh = chip ? ' width="120" height="18"' : ' width="480" height="72"';
    return (
      '<picture' + (chip ? '' : ' class="fuel-banner__picture"') + '>' +
        '<source type="image/svg+xml" srcset="' + L.svg + '">' +
        '<img class="' + cls + '" src="' + L.png + '" alt="' + escapeHtml(L.alt) + '"' +
          wh + ' loading="lazy" decoding="async">' +
      '</picture>'
    );
  }

  /**
   * @param {object} opts
   * @param {string} opts.combustible
   * @param {string} [opts.matricula]
   * @param {string} [opts.tipo]
   * @param {string} [opts.cliente]
   * @param {'cliente'|'staff'} [opts.audience]
   * @param {'turno'|'maestro'} [opts.variant]
   * @param {boolean} [opts.lockedByMaestro]
   */
  function buildFuelBanner(opts) {
    const audience = opts.audience || 'cliente';
    const variant = opts.variant || 'turno';
    const locked = !!(opts.lockedByMaestro || variant === 'maestro');
    const { kind, label, colorCode } = resolveKind(opts.combustible);
    if (!kind) {
      return {
        show: false, kind: 'empty', label: '', colorCode: '', cssMod: '', bg: '',
        subline: '', logoSvg: '', logoPng: '', logoAlt: '', html: '',
      };
    }

    let cssMod = 'fuel-banner--unknown';
    let bg = 'transparent';
    if (kind === 'avgas') { cssMod = 'fuel-banner--avgas'; bg = AVGAS_BG; }
    else if (kind === 'jet') { cssMod = 'fuel-banner--jet'; bg = JET_BG; }

    const logo = LOGO[kind] || null;
    const mat = String(opts.matricula || '').trim();
    const tipo = cleanTipo(opts.tipo, audience === 'cliente');
    const cli = audience === 'staff' ? String(opts.cliente || '').trim() : '';
    const codeTxt = colorCode ? ('Código de color internacional: ' + colorCode) : '';

    let subline = '';
    if (locked && mat) {
      const tip = tipo ? (' (' + tipo + ')') : '';
      const head = audience === 'staff'
        ? ('Grado bloqueado por el maestro para ' + mat + tip)
        : ('Grado según matrícula ' + mat + tip);
      subline = codeTxt ? (head + ' · ' + codeTxt) : head;
    } else {
      const parts = [];
      if (mat) parts.push(mat);
      if (tipo) parts.push(tipo);
      if (cli) parts.push(cli);
      if (codeTxt) parts.push(codeTxt);
      subline = parts.join(' · ');
    }

    return {
      show: true,
      kind,
      label,
      colorCode,
      cssMod,
      bg,
      subline,
      logoSvg: logo ? logo.svg : '',
      logoPng: logo ? logo.png : '',
      logoAlt: logo ? logo.alt : label,
      html: '', // filled by renderFuelBanner
    };
  }

  function renderFuelBanner(opts) {
    const b = buildFuelBanner(opts || {});
    if (!b.show) return '';
    const sticky = opts && opts.sticky ? ' fuel-banner--sticky' : '';
    const logoClass = b.logoSvg ? ' fuel-banner--logo' : '';
    const styleBg = b.logoSvg ? '' : (' style="background:' + b.bg + '"');

    let body;
    if (b.logoSvg) {
      body = logoMarkup(b.kind);
    } else {
      body = (
        '<div class="fuel-banner__title">' +
          '<span class="fuel-banner__icon" aria-hidden="true">⛽</span>' +
          '<span class="fuel-banner__grade">' + escapeHtml(b.label) + '</span>' +
        '</div>'
      );
    }

    return (
      '<div class="fuel-banner ' + b.cssMod + logoClass + sticky + '" role="status"' + styleBg + '>' +
        body +
        (b.subline ? ('<p class="fuel-banner__sub">' + escapeHtml(b.subline) + '</p>') : '') +
      '</div>'
    );
  }

  /** Chip compacto para filas de tabla — logo pequeño si hay. */
  function fuelChipHtml(combustible) {
    const { kind, label } = resolveKind(combustible);
    if (!kind) return '';
    if (kind === 'jet' || kind === 'avgas') {
      const chipClass = kind === 'jet' ? 'chip--jet' : 'chip--avgas';
      const alt = LOGO[kind].alt;
      return (
        '<span class="chip chip--logo ' + chipClass + '" title="' + escapeHtml(alt) + '">' +
          logoMarkup(kind, { chip: true }) +
        '</span>'
      );
    }
    return '<span class="chip">' + escapeHtml(label) + '</span>';
  }

  /** Chip con label completo (compat con paneles existentes). */
  function fuelChipFullHtml(combustible) {
    const { kind, label } = resolveKind(combustible);
    if (!kind) return '';
    if (kind === 'jet' || kind === 'avgas') {
      return fuelChipHtml(combustible);
    }
    return combustible ? ('<span class="chip">' + escapeHtml(combustible) + '</span>') : '';
  }

  global.FuelBanner = {
    build: buildFuelBanner,
    render: renderFuelBanner,
    chip: fuelChipHtml,
    chipFull: fuelChipFullHtml,
    resolve: resolveKind,
    escapeHtml,
  };
})(typeof window !== 'undefined' ? window : globalThis);

File: app/fuel_banner.py
"""Banner de grado de combustible (código de color internacional).

AVGAS 100LL → rojo · JET A-1 → negro.
Usado en plantillas Jinja y como fuente de verdad para el helper JS.
Logos oficiales en /static/img/fuel/ (SVG preferido; PNG para email / fallback).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

Audience = Literal["cliente", "staff"]
Variant = Literal["turno", "maestro"]

# Rutas estáticas relativas (templates web / JS)
_LOGO_SVG = {
    "jet": "/static/img/fuel/jet-a1.svg",
    "avgas": "/static/img/fuel/avgas-100ll.svg",
}
_LOGO_PNG = {
    "jet": "/static/img/fuel/jet-a1-oficial.png",
    "avgas": "/static/img/fuel/avgas-100ll-oficial.png",
}

# Alt accesible (espacio en 100 LL según branding oficial)
_LOGO_ALT = {
    "jet": "JET A-1",
    "avgas": "AVGAS 100 LL",
}


def _normalize_grado(value: str) -> str:
    """Copia ligera de matricula.normalize_grado (sin importar SQLAlchemy)."""
    s = (value or "").upper().strip()
    s = s.replace("-", " ").replace("_", " ")
    s = " ".join(s.split())
    compact = s.replace(" ", "")
    if compact in {"JETA1", "JETA-1"} or "JET" in s or "AEROKEROSENE" in s:
        return "JET A-1"
    if "AVGAS" in s or "100LL" in s or "100 LL" in s:
        return "AVGAS 100LL"
    return s


@dataclass(frozen=True)
class FuelBanner:
    """Datos listos para renderizar el banner."""

    kind: str  # avgas | jet | unknown | empty
    label: str
    color_code: str  # ROJO | NEGRO | ""
    css_mod: str  # fuel-banner--avgas | --jet | --unknown
    bg: str
    subline: str
    show: bool
    logo_svg: str = ""  # path relativo o vacío
    logo_png: str = ""
    logo_alt: str = ""


_AVGAS_BG = "#C0392B"
_JET_BG = "#1C1C1C"
_UNKNOWN_BG = "transparent"


def _clean_tipo(tipo: str | None, *, for_cliente: bool) -> str:
    t = (tipo or "").strip()
    if not t:
        return ""
    if for_cliente and t.upper() in {"S/D", "SD", "N/D", "ND", "-", "—"}:
        return ""
    return t


def _clean_part(value: str | None) -> str:
    return (value or "").strip()


def resolve_fuel_kind(combustible: str | None) -> tuple[str, str, str]:
    """→ (kind, label, color_code). kind empty si no hay grado."""
    raw = (combustible or "").strip()
    if not raw:
        return ("", "", "")
    g = _normalize_grado(raw)
    if g == "AVGAS 100LL" or "AVGAS" in g or "100LL" in g.replace(" ", ""):
        return ("avgas", "AVGAS 100LL", "ROJO")
    if g == "JET A-1" or "JET" in g:
        return ("jet", "JET A-1", "NEGRO")
    return ("unknown", g or raw, "")


def build_fuel_banner(
    *,
    combustible: str | None,
    matricula: str | None = None,
    tipo: str | None = None,
    cliente: str | None = None,
    audience: Audience = "cliente",
    variant: Variant = "turno",
    locked_by_maestro: bool = False,
) -> FuelBanner:
    """Arma título + sublínea según handoff UX."""
    kind, label, color_code = resolve_fuel_kind(combustible)
    if not kind:
        return FuelBanner(
            kind="empty",
            label="",
            color_code="",
            css_mod="",
            bg="",
            subline="",
            show=False,
        )

    if kind == "avgas":
        css_mod, bg = "fuel-banner--avgas", _AVGAS_BG
    elif kind == "jet":
        css_mod, bg = "fuel-banner--jet", _JET_BG
    else:
        css_mod, bg = "fuel-banner--unknown", _UNKNOWN_BG

    logo_svg = _LOGO_SVG.get(kind, "")
    logo_png = _LOGO_PNG.get(kind, "")
    logo_alt = _LOGO_ALT.get(kind, label)

    mat = _clean_part(matricula)
    tipo_clean = _clean_tipo(tipo, for_cliente=(audience == "cliente"))
    cli = _clean_part(cliente) if audience == "staff" else ""
    code_txt = f"Código de color internacional: {color_code}" if color_code else ""

    use_maestro = locked_by_maestro or variant == "maestro"
    if use_maestro and mat:
        if audience == "staff":
            tip = f" ({tipo_clean})" if tipo_clean else ""
            head = f"Grado bloqueado por el maestro para {mat}{tip}"
        else:
            tip = f" ({tipo_clean})" if tipo_clean else ""
            head = f"Grado según matrícula {mat}{tip}"
        parts = [head]
        if code_txt:
            parts.append(code_txt)
        subline = " · ".join(parts)
    else:
        parts: list[str] = []
        if mat:
            parts.append(mat)
        if tipo_clean:
            parts.append(tipo_clean)
        if cli:
            parts.append(cli)
        if code_txt:
            parts.append(code_txt)
        subline = " · ".join(parts) if parts else code_txt

    return FuelBanner(
        kind=kind,
        label=label,
        color_code=color_code,
        css_mod=css_mod,
        bg=bg,
        subline=subline,
        show=True,
        logo_svg=logo_svg,
        logo_png=logo_png,
        logo_alt=logo_alt,
    )


def fuel_banner_dict(**kwargs) -> dict:
    """Dict para JSON / JS (misma forma que el helper de front)."""
    b = build_fuel_banner(**kwargs)
    return {
        "show": b.show,
        "kind": b.kind,
        "label": b.label,
        "color_code": b.color_code,
        "css_mod": b.css_mod,
        "bg": b.bg,
        "subline": b.subline,
        "logo_svg": b.logo_svg,
        "logo_png": b.logo_png,
        "logo_alt": b.logo_alt,
    }
